include/caffe/datum_DB.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef CAFFE_DATUMDB_HPP
+#define CAFFE_DATUMDB_HPP
+
+#include <string>
+
+#include "leveldb/db.h"
+#include "lmdb.h"
+
+#include "caffe/common.hpp"
+#include "caffe/datum_DB_factory.hpp"
+#include "caffe/proto/caffe.pb.h"
+
+namespace caffe {
+
+class DatumDB {
+ public:
+  explicit DatumDB(const DatumDBParameter& param)
+    : param_(param),
+      is_opened_(new bool(false)) {}
+  virtual ~DatumDB() {
+    if (is_opened_.unique()) {
+      DatumDBRegistry::RemoveSource(param_.source());
+    }
+  }
+
+  class Generator {
```

Comment:
That feels confusing and underhanded; why not refer instead of copying?

Does this support multiple simultaneous cursors? If so, the interface should reflect that by implementing the cursor methods (`Next`, `Reset`, &c) in the cursor or iterator subclass (or with a cursor or iterator as arguments, which amounts to the same thing). If not, there's no need for a nested class.